### PR TITLE
[perf] FSDP: speed up no_sync and test communication volume

### PR DESCRIPTION
**Pull request by myleott**
This reduces communication when inside a ``no_sync`` context.

In particular, the root instance's params do not need to be freed/regathered each time when accumulating gradients. This speedups up `--update-freq` in fairseq, particularly for small and medium-sized models.

Also added a test to confirm expected communication volume in this case, and moved TestNoSync out of test_fsdp.py into a new file: test_fsdp_no_sync.py

### CODE REVIEW

File: tests/nn/data_parallel/test_fsdp_no_sync.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+            loss1 = model.module.get_loss(batch1, output)
+            loss1.backward()
+        output = model(*batch2)
+        loss2 = model.module.get_loss(batch2, output)
+        loss2.backward()
+        accumulated_loss = loss1 + loss2
+        accumulated_grads = [p.grad.detach().clone() for p in model.parameters()]
+
+        torch.testing.assert_allclose(ref_loss, accumulated_loss)
+        assert objects_are_equal(ref_grads, accumulated_grads, raise_exception=True)
```
Comment (myleott) [inline tests/nn/data_parallel/test_fsdp_no_sync.py:94]: This is all copied/pasted out of test_fsdp.py

File: tests/nn/data_parallel/test_fsdp_no_sync.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+
+                assert mock_all_gather.call_count == 1
+                assert mock_reduce_scatter.call_count == 0
+
+                output = model(*batch)
+                loss = model.module.get_loss(batch, output)
+                loss.backward()
+
+                assert mock_all_gather.call_count == 1
+                assert mock_reduce_scatter.call_count == 1
```
Comment (myleott) [inline tests/nn/data_parallel/test_fsdp_no_sync.py:127]: This is the new test

File: tests/nn/data_parallel/test_fsdp_no_sync.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+import torch
+
+from fairscale.nn.data_parallel import FullyShardedDataParallel
+from fairscale.utils.testing import DummyProcessGroup, objects_are_equal
+
+from .test_fsdp import DistributedTest, NestedWrappedModule, spawn_and_init
+
+
+class TestNoSync(DistributedTest):
```
Comment (myleott) [inline tests/nn/data_parallel/test_fsdp_no_sync.py:18]: This test is copy/pasted, no need to review

File: tests/nn/data_parallel/test_fsdp_no_sync.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+        loss2 = model.module.get_loss(batch2, output)
+        loss2.backward()
+        accumulated_loss = loss1 + loss2
+        accumulated_grads = [p.grad.detach().clone() for p in model.parameters()]
+
+        torch.testing.assert_allclose(ref_loss, accumulated_loss)
+        assert objects_are_equal(ref_grads, accumulated_grads, raise_exception=True)
+
+
+class TestNoSyncCommunication(DistributedTest):
```
Comment (myleott) [inline tests/nn/data_parallel/test_fsdp_no_sync.py:97]: This is the new test

**Review by sshleifer — APPROVED** [review]
LGTM, Nice Test!

**Comment by min-xu-ai** [comment]
is this still a draft?

**Comment by myleott** [comment]
Yes, some tests are failing (real failures), still digging
